numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
To follow up on my comment about the Python bug, it occurs to me that that is actually not really an argument either way, since our mixin class could also include a fix. Still, it's an argument in favor of @njsmith's proposal for option 1, since it means that `__array_ufunc__` only needs to worry about NumPy's dispatch rules, not Python's subtly different ones.

@mhvk I am finally remembering some of the details of our previous long discussion. I agree with @njsmith we want to avoid complicated / hard-to-understand interaction between `__array_ufunc__` and Python's binary operator dispatch, and raising TypeError immediately if any ufunc arguments set `__array_ufunc__ = None` will help. In particular, there are two secondary design goals that we would like to achieve here:

1. Make it impossible (or at least very difficult) for `np.add(x, y)` and `x + y` to return different results.
2. Allow classes to opt out of binary arithmetic in the numpy/`__array_ufunc__` way and implement arithmetic in the normal python way, e.g., by implementing `__add__`/`__radd__`/`__iadd__`.

The current implementation allows for indicating (2) by setting `__array_ufunc__ = None`. This guarantees that no *single class* can choose to implement different semantics for `np.add(x, y)` and `x + y`, but it doesn't stop *another class* that overloads `__array_ufunc__` from recognizing the first class and implementing the ufunc. To give a concrete example, which is possible with the current mixin and override implementation, consider these two classes:
```
class A:
    __numpy_ufunc__ = None
    def __add__(self, other):
        ...
    def __radd__(self, other):
        ...

class B(np.lib.NDOperatorsMixin):
    def __numpy_ufunc__(self, ufunc, method, *inputs, **kwargs):
        ...
```
Both `A() + B()` and `B() + A()` will use `A`'s implementation for `+` (either `A.__add__` or `A.__radd__`. But `np.add(A(), B())` and `np.add(B(), A())` will instead rely on `B.__numpy_ufunc__`. In principle, these could give inconsistent results.

On the other hand, if we update our override rules to immediately raise `TypeError` if any arguments to a ufunc set `__numpy_ufunc__ = None`, `np.add(A(), B())` will error instead. This is still a reasonable restriction:`A` opted out of ufuncs, so it can't be used in them, even by another class.